### DISCUSSION TITLE
Can I update the token of a cloned repo?

### DISCUSSION THREAD

**Discussion by aleks-ivanov** [community/community#22434] [New to GitHub]
<p>A long while back I cloned a repo I own with a PAT.</p>
<p>Recently, since the GitHub PAT algorithm update, I changed my PATs, so now the cloned repo can’t auth properly and therefor I can’t push to origin.</p>
<p>Is there a way to update the token of the cloned repo, so I don’t have to remove the local clone and clone it all over again ?</p>

**Comment by byrneh** [comment]
<p>Hi <a class="mention" href="/aleks-ivanov">@aleks-ivanov</a>, personal access tokens are associated with user accounts not your repository.<br>
If you have not rẻ-entered your new PAT ata prinoy then most likely you are using a Git client where you Credential is cached.<br>
You can safely delete your old cached Credential, if this is the case and you should get prompted to enter a new one in tbe next authenticated git command</p>
<p>For Windows example see here…</p><aside class="quote quote-modified" data-post="2" data-topic="163304">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/how-to-start-using-a-personal-access-token-to-authenticate-git-on-windows/163304/2">How to start using a personal access token to authenticate git on windows</a> <a class="badge-wrapper  bullet" href="/c/github-help/how-to-use-git-and-github/21"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #ffd1ac;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Ask and answer questions regarding Git and GitHub.">How to use Git and GitHub</span></a>
  </div>
  <blockquote>
    If you are still using an existing cached password for you login you will need to remove it, the below will help, hopefully slight_smile 
Launch ‘Credential Manager’ on your Windows device. 
Switch to tab Windows Credentials (it default displayed tab Web Credentials on my device). 
A list of locally cached credentials are then shown 
Find an entry starting something like git:htttps//login@github.com (this will be an exist cached entry for your username/password.  You can safely delete this by …
  </blockquote>
</aside>

**Comment by aleks-ivanov** [comment]
<p>I never use cached credentials, I always use per cloned repo authentication like this:</p>
<pre><code class="lang-auto">git clone https://token:jnibf29833fibf293gfb2987fg42gb0vf@github.com/aleks-ivanov/project.git
</code></pre>

**Comment by airtower-luna** [comment]
<p>That doesn’t look like one of the current PATs (which should start with <code>ghp_</code>). However, if that is an old token, you should revoke it now after sharing it on a public forum. :warning:</p>

**Comment by aleks-ivanov** [comment]
<p>Not an actual token of course, just a random smash of the keyboard as example</p>

**Comment by byrneh** [comment]
<p>If your token is entered correctly but still not working you coul check it has the correct repo scope enabled.<br>
I don’t personally specify in git like you are doing here but have you tried format<br>
username:token@…</p>

**Comment by aleks-ivanov** [comment, marked as answer]
<p>The solution to my original problem is the following:</p>
<pre><code class="lang-auto">$newGitHubPAT = "ghp_..."
$newOrigin = "https://token:$newGitHubPAT@github.com/owner/repo.git"
git remote remove origin # origin is the remote repo
git remote add origin $newOrigin
</code></pre>
<p>Basically you just remove the old origin that referenced the remote repository with the old token and add the origin with the new PAT.</p>

**Reply by asteve12335** [reply]
I propose you don't save your info in some variables
Just use it when needed

**Reply by YassineRKB** [reply]
the problem is when you are pushing multiple changes in rather small time frame it gets tedious to insert your username and token each time compared to storing token in origin

**Reply by YassineRKB** [reply]
thats what brought me here, looking for a better solution.

**Reply by airtower-luna** [reply]
@YassineRKB, see [Git Tools - Credential Storage](https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage). If you just don't want to type/copy from password manager all the time but without a plaintext token on disk, the `cache` method might be the right one.

**Comment by airtower-luna** [comment]
<p>This works, but is potentially dangerous because the token ends up stored on your disk unencrypted. Anyone with access to the filesystem could copy it.</p>

**Comment by aleks-ivanov** [comment]
<p>I agree, but it is really convenient :slightly_smiling_face:</p>

**Comment by Kemuntofaith** [comment]
hello. am not answering your question but asking my own. i cloned a repo using a pat a while back but the pat expired. i generated a new pat but every time i try to push to that repo, i am asked for my password. it is very annoying to copy and paste the pat every time i want to push my work to git. how do i deal with this please. id like to be able to push my work as efficiently as i did before.

**Reply by airtower-luna** [reply]
A new question should be a new thread. :wink: To minimize off-topic, here just a link with documentation on how to store credentials: https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage The `cache` method plus password manager is usually a good solution.

**Comment by codealchemist** [comment]
We can avoid exposing the PAT by removing and adding back the remote.
```
git remote remove origin
git remote add origin [origin]
```

Then when we try to push we are asked for username and password like the first time, and we can paste the PAT as the password, keeping it safe.

**Reply by mesirhc** [reply]
> asked for username and password like the first time,

**Exactly** what I've been looking for!

What do you mean by `[origin]` in:
`    git remote add origin [origin]`
??? https://github.com/repo_owner/repo_name.git ...or ?

Sorry if it seems obvious to you, but I have spent hours following dead-ends on the internet: I want this to work!
ChrisM

**Reply by airtower-luna** [reply]
Exactly, replace `[origin]` with the URL you want to use to fetch from the repository. I also want to point out that instead of remove/add it's better to use [`git remote set-url`](https://git-scm.com/docs/git-remote#Documentation/git-remote.txt-emset-urlem). It's only one step *and* doesn't lose other remote settings.

**Reply by mesirhc** [reply]
I've spent four hours ploughing through suggestions on the internet - this thread was one of the first half dozen, but I stopped reading when I saw the accepted answer.
Second-visit, I read to the end of the thread and saw @codealchemist pointing in the right direction;
then @airtower-luna absolutely nails it with a one-liner!

Thanks so much to both! 
(Airtower: this is the second time you have de-mystified GitHub for me, so extra thanks).

ChrisM

**Reply by codealchemist** [reply]
A pleasure to help @mesirhc 🙌 
Loved the `set-url`, thanks @airtower-luna 🎉
